sql: provide table comments for virtual tables

hueypark opened the pull request:
Fixes #32964
Fixes #35581

Release note (sql change): CockroachDB now provides usable comments
with optional documentation URLs for the virtual tables in
`pg_catalog`, `information_schema` and `crdb_internal`. Use `SHOW
TABLES [FROM ...] WITH COMMENT` to read. Note that `crdb_internal`
tables remain an experimental feature subject to change without
notice.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/34764)
<!-- Reviewable:end -->

knz commented on pkg/sql/crdb_internal.go:
```diff
@@ -197,7 +197,36 @@ CREATE TABLE crdb_internal.tables (
+			e := vEntries[virtSchemaName]
+			for _, tName := range e.orderedDefNames {
+				table := e.defs[tName].desc
+				if err := addRow(
```
make a single function that takes the desc as argument and does the common thing between this and below

knz commented on pkg/sql/crdb_internal.go:
```diff
@@ -644,6 +673,7 @@ CREATE TABLE crdb_internal.session_trace (
   message     STRING NOT NULL,     -- The logged message.
   age         INTERVAL NOT NULL    -- The age of this message relative to the beginning of the trace.
 )`,
+	comment: "latest trace collected on this session (via SET TRACING={ON/OFF}).",
```
nit/style: when we opt out of full sentences (no capital at start) we also opt out of the final period. (Same as cluster settings)

Conversely you could opt into the capitals.

knz commented on pkg/sql/crdb_internal.go:
```diff
@@ -736,7 +768,8 @@ func (p *planner) makeSessionsRequest(ctx context.Context) serverpb.ListSessions
 var crdbInternalLocalQueriesTable = virtualSchemaTable{
-	schema: fmt.Sprintf(queriesSchemaPattern, "node_queries"),
+	schema:  fmt.Sprintf(queriesSchemaPattern, "node_queries"),
+	comment: "the list of running queries on the current node. The results are dependent on the current user.",
```
inconsistent articles (either use "the" or not at start, but do the same consistently)

knz commented on pkg/sql/crdb_internal.go:
```diff
@@ -960,7 +996,7 @@ var crdbInternalLocalMetricsTable = virtualSchemaTable{
   value							 FLOAT NOT NULL    -- value of the metric
 )`,
-
+	comment: "snapshot of the metrics on the current node.",
```
not sure "snapshot" is necessary. I think it's implied for all the tables.

knz commented on pkg/sql/crdb_internal.go:
```diff
@@ -1038,6 +1075,7 @@ CREATE TABLE crdb_internal.create_statements (
   validate_statements STRING[] NOT NULL
 )
 `,
+	comment: "exposes the CREATE TABLE/CREATE VIEW statements.",
```
"the CREATE statements". No need to be specific, also your current phrasing would already be inaccurate.

hueypark reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34764)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/sql/crdb_internal.go, line 209 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34764#-LYLP7qpWDpV1gw3anCI-r1-209:-LZjGUS91kSNxqzXnYaf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cde54faa0d91212f9eec4fa1259d4ba23ae34a3/pkg/sql/crdb_internal.go#L209)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

make a single function that takes the desc as argument and does the common thing between this and below
</blockquote></details>

Done.
___
*[pkg/sql/crdb_internal.go, line 676 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34764#-LYLP7qpWDpV1gw3anCI-r1-676:-LZjH-Fh0RkZ5teHR4yZ:bbw8fsc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cde54faa0d91212f9eec4fa1259d4ba23ae34a3/pkg/sql/crdb_internal.go#L676)):*
> nit/style: when we opt out of full sentences (no capital at start) we also opt out of the final period. (Same as cluster settings)
I choose to opt out of the final period.
___
*[pkg/sql/crdb_internal.go, line 772 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34764#-LYLP7qpWDpV1gw3anCI-r1-772:-LZjH1nEBXzjw92kwkm3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cde54faa0d91212f9eec4fa1259d4ba23ae34a3/pkg/sql/crdb_internal.go#L772)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

inconsistent articles (either use "the" or not at start, but do the same consistently)
</blockquote></details>

Done.
___
*[pkg/sql/crdb_internal.go, line 999 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34764#-LYLP7qpWDpV1gw3anCI-r1-999:-LZjHBzK45_hBVnv3VZc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cde54faa0d91212f9eec4fa1259d4ba23ae34a3/pkg/sql/crdb_internal.go#L999)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

not sure "snapshot" is necessary. I think it's implied for all the tables.
</blockquote></details>

Done.
___
*[pkg/sql/crdb_internal.go, line 1078 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34764#-LYLP7qpWDpV1gw3anCI-r1-1078:-LZjHRO1Ai_ne5QahAJd:be8cqz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cde54faa0d91212f9eec4fa1259d4ba23ae34a3/pkg/sql/crdb_internal.go#L1078)):*
> "the CREATE statements". No need to be specific, also your current phrasing would already be inaccurate.
I deleted the unnecessary comment.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
Thank you for your changes. I have some more suggestions below.

> Reviewed 2 of 8 files at r1, 6 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34764#-:-LZq-gX93GK5zvsyZ9X3:b-4k7npp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @hueypark)

___
*[pkg/server/admin.go, line 1494 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34764#-LZpyqhJDViKBqmwgdVk:-LZpyqhJDViKBqmwgdVl:b-pxt9jn) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ad40ed2abee91fc9fbe26e8a4d7c927698a80d/pkg/server/admin.go#L1494)):*
> ```Go
> 		dbName := (*string)(row[2].(*tree.DString))
> 
> 		if sqlbase.IsVirtualTable(sqlbase.ID(tableID)) {
> ```

Thanks for noticing this. However I think this change belongs to a different PR. Can you extract it from this PR and file a different PR for this? Thank you
___
*[pkg/sql/crdb_internal.go, line 209 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34764#-LYLP7qpWDpV1gw3anCI-r1-209:-LZpz-7r-ZAI_7mNfaw4:bzi573l) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cde54faa0d91212f9eec4fa1259d4ba23ae34a3/pkg/sql/crdb_internal.go#L209)):*
<details><summary><i>Previously, hueypark (Jaewan Park) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

This will allocate an array, expand it and discard it for every row. Maybe it's a bit memory-heavy and inefficient? Consider this instead:

```go
createTableRow := func(table *sqlbase.TableDescriptor, dbName string) error {
   ...
       return addRow(...)
}
```

(i.e. use a closure)

then below

```go
if err := createTableRow(table, dbName); err != nil {
   return err
}
```
___
*[pkg/sql/crdb_internal.go, line 676 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34764#-LYLP7qpWDpV1gw3anCI-r1-676:-LZpzYyv7kNGUz-NhKwf:bbdbffo) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cde54faa0d91212f9eec4fa1259d4ba23ae34a3/pkg/sql/crdb_internal.go#L676)):*
<details><summary><i>Previously, hueypark (Jaewan Park) wrote…</i></summary><blockquote>

> nit/style: when we opt out of full sentences (no capital at start) we also opt out of the final period. (Same as cluster settings)
I choose to opt out of the final period.
</blockquote></details>

I still see the period here :)

this is a period: `.`

for example `"latest trace collected on this session."` -> `"latest trace collected on this session"`
___
*[pkg/sql/show_tables.go, line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34764#-LZpzpbu3YXc5QKwAoQ2:-LZpzpbu3YXc5QKwAoQ3:bpn0djt) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ad40ed2abee91fc9fbe26e8a4d7c927698a80d/pkg/sql/show_tables.go#L52)):*
> ```Go
> SELECT
> 	i.table_name,
> 	obj_description(t.table_id::REGCLASS) AS comment
> ```

This seems to be better because it's shorter, however it's objectively worse for performance: the query optimizer cannot "See through" `obj_description` and see the join performed within, and cannot work to simplify the query. The result is that the comment table will be queried again for every table.

However I also understand that `system.comment` does not contain the virtual table comments.

Maybe the right way to do this is to add a new virtual table in `information_schema` (`information_schema.comments`?) that contains all the data from `system.comments` + the data in the `.comment` field for virtual tables. 

This way you can preserve the join here, simply replacing `system.comments` by `information_schema.comments`.


<!-- Sent from Reviewable.io -->

hueypark committed:
sql: provide table comments on virtual tables

This patch adds a `comment` to `virtualSchemaTable` and exposes it
through a new virtual table `crdb_internal.predefined_comments`.

This is then used by `pg_catalog.pg_description` to expose comments on
virtual tables alongside those of real tables. This way
`obj_description()` works on virtual tables too.

Finally it modifies `SHOW TABLES WITH COMMENT` to use `pg_description`
and expose the table comments.

Release note (sql change): CockroachDB now provides usable comments
with optional documentation URLs for the virtual tables in
`pg_catalog`, `information_schema` and `crdb_internal`. Use `SHOW
TABLES [FROM ...] WITH COMMENT` to read. Note that `crdb_internal`
tables remain an experimental feature subject to change without
notice.

hueypark reviewed (commented):
Thank you for your reviews. I changed the implementation based on your suggestion(add `information_schema.comments`).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34764#-:-L_3YM_-DXCw7VFVl_Io:bd3tvbm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/sql/crdb_internal.go, line 209 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34764#-LYLP7qpWDpV1gw3anCI-r1-209:-L_3Yif5CpXOO2mxBcNd:b-kmpxpc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cde54faa0d91212f9eec4fa1259d4ba23ae34a3/pkg/sql/crdb_internal.go#L209)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This will allocate an array, expand it and discard it for every row. Maybe it's a bit memory-heavy and inefficient? Consider this instead:

```go
createTableRow := func(table *sqlbase.TableDescriptor, dbName string) error {
   ...
       return addRow(...)
}
```

(i.e. use a closure)

then below

```go
if err := createTableRow(table, dbName); err != nil {
   return err
}
```
</blockquote></details>

This function is now obsolete.
___
*[pkg/sql/crdb_internal.go, line 676 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34764#-LYLP7qpWDpV1gw3anCI-r1-676:-L_3a4_1525y1uG1k5ta:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cde54faa0d91212f9eec4fa1259d4ba23ae34a3/pkg/sql/crdb_internal.go#L676)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I still see the period here :)

this is a period: `.`

for example `"latest trace collected on this session."` -> `"latest trace collected on this session"`
</blockquote></details>

Done.
___
*[pkg/sql/show_tables.go, line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34764#-LZpzpbu3YXc5QKwAoQ2:-L_3a8RF5IpgdGlYx5ur:b-t07l50) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ad40ed2abee91fc9fbe26e8a4d7c927698a80d/pkg/sql/show_tables.go#L52)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This seems to be better because it's shorter, however it's objectively worse for performance: the query optimizer cannot "See through" `obj_description` and see the join performed within, and cannot work to simplify the query. The result is that the comment table will be queried again for every table.

However I also understand that `system.comment` does not contain the virtual table comments.

Maybe the right way to do this is to add a new virtual table in `information_schema` (`information_schema.comments`?) that contains all the data from `system.comments` + the data in the `.comment` field for virtual tables. 

This way you can preserve the join here, simply replacing `system.comments` by `information_schema.comments`.
</blockquote></details>

Done.
And I've added an id column to the `information_schema.tables` table to eliminate the unnecessary JOIN.
___
*[pkg/server/admin.go, line 1494 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34764#-LZpyqhJDViKBqmwgdVk:-L_3ajuYA6YEDNlPc5mB:bgf9rux) ([raw file](https://github.com/cockroachdb/cockroach/blob/75ad40ed2abee91fc9fbe26e8a4d7c927698a80d/pkg/server/admin.go#L1494)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Thanks for noticing this. However I think this change belongs to a different PR. Can you extract it from this PR and file a different PR for this? Thank you
</blockquote></details>

Done.
But, This was necessary to pass the test If I do not use `information_schema.comments` approach.


<!-- Sent from Reviewable.io -->

knz commented:
Thank you for your changes. All it all it was a good improvement.

I made a few last minute changes, including fixing `col_description` and `obj_description`. I also decided against adding `table_id` to `information_schema.tables`, because the risk of problem with 3rd party tools was too important. Instead I made `show tables with comment` use `pg_description` and aded `crdb_internal.predefined_comments` to support that.

Again thank you for your efforts!

bors r+
